gems/aws-sdk-s3/lib/aws-sdk-s3/file_downloader.rb

```diff
@@ -152,49 +117,58 @@ def multithreaded_get_by_parts(n_parts, total_size, etag)
       def download_in_threads(pending, total_size)
         threads = []
         progress = MultipartProgress.new(pending, total_size, @progress_callback) if @progress_callback
+        @temp_path = "#{@path}.s3tmp.#{SecureRandom.alphanumeric(8)}"
         @thread_count.times do
           thread = Thread.new do
             begin
-              while part = pending.shift
+              while (part = pending.shift)
                 if progress
                   part.params[:on_chunk_received] =
                     proc do |_chunk, bytes, total|
                       progress.call(part.part_number, bytes, total)
                     end
                 end
                 resp = @client.get_object(part.params)
-                write(resp)
+                range = extract_range(resp.content_range)
+                validate_range(range, part.params[:range]) if part.params[:range]
+                write(resp.body, range)
                 if @on_checksum_validated && resp.checksum_validated
                   @on_checksum_validated.call(resp.checksum_validated, resp)
                 end
               end
               nil
-            rescue => error
-              # keep other threads from downloading other parts
-              pending.clear!
-              raise error
+            rescue StandardError => e
+              pending.clear! # keep other threads from downloading other parts
+              raise e
             end
           end
           threads << thread
         end
         threads.map(&:value).compact
+        File.rename(@temp_path, @path)
```

Comment:
I think push this up to the download() method as the last thing that happens, next to where we do the ensure for clean up. This method is for downloading parts.